design/0003-isolated-integration-tests.md

```diff
@@ -0,0 +1,324 @@
+- Have a more discoverable UX for running all or one test
+- Less side effects when running the tests (and less footguns)
+- Easier to extend for more complex integration tests
+
+The two main components that the following subsections talk about are the code
+component and the script component.
+
+The code component is comprised of the Go files in the `test/` directory and
+have direct interaction with `go test`. The Go test files located here are
+responsible for testing specific features of Linkerd such as service profiles,
+tap, edges, and more--they are not the actual integration tests. It is
+important to note that these tests assume the cluster state is already set,
+and it should stay that way. This allows developers to run `go test` and
+specify the cluster context, namespace, and other knobs that may be important
+for properly testing certain cases.
+
+The script component is comprised of the `test-run` and `_test-run.sh`
+scripts. These are the actual integration tests. Each integration test is
+comprised of cluster setup and some combination of the test files from the
+code component.
+
+The separation here is important to note for this change: the integration
+tests are contained within scripts and responsible for managing the cluster
+state and running some combination of Go test files.
+
+Therefore, these changes will be mostly script based. There is some possible
+tooling that will allow managing cluster state through code (discussed in `#
+Future  possibilities`) but the ecosystem is currently not in a state where we
+can avoid breaking out to the shell at some point for configuring something
+about the environment.
+
+### UX
+
+Running the integration tests will be done through new scripts that are
+responsible for the setup and teardown of their testing environments.
+
+This is different from the current situation in a few ways. First, there is
+only one self-contained script that runs the integration test suit; it runs
+all the tests and does not provide a way to select specific ones. If one does
+wish to run a specific test, they must export the internal test script into
+their shell environment.
+
+Second, the environment setup check must be run before each integration test.
+This is automatic in the self-contained script, but is a required manual step
+for running a specific one.
+
+It will be important to document how developers can run all or one of the
+tests, so providing clear instructions in the Test markdown file, as well as a
+`--help` flag for the scripts will be included.
+
+Before: Running all tests:
+
+```
+$ bin/test-run $PWD/target/cli/darwin/linkerd
+```
+
+Before: Running specific test:
+
+```
+$ . bin/_test-run.sh
+$ init_test_run $PWD/target/cli/darwin/linkerd
+$ deep_integration_tests
+```
+
+After: Running all tests:
+
+```
+$ bin/tests $PWD/target/cli/darwin/linkerd
+```
+
+After: Running specific test:
+
+```
+$ bin/deep-tests $PWD/target/cli/darwin/linkerd
```

Comment:
I think this RFC should address the CLI argument. I never understood why we have to do this manually and it can't be done by the test harness...

Reply:
I usually have multiple CLI versions in my directory. If I'm testing something I just changed, then I would pass the `target/cli/..` binary. If I'm testing a version that I pulled with `docker-pull-binaries`, I would want to specify which version from `target/release/linkerd2-cli-..`.

I think the test harness handles this particularly well in that you can run the `test/` test files or integration tests with a specific version of the CLI.

If we moved this into the test harness, the developer would still need some way to specify that they want to run the tests with the CLI they just built, or a CLI that they just pulled from a PR or past release.

Reply:
I understand that it's useful to have the ability to use a custom CLI binary, but I think the vast majority of the time this is an unnecessary hurdle. I can't tell you how many times I've accidentally tested a stale binary because I was missing a rebuild...

I think about it like this... If I was running `go test ...`, it would handle building everything it needs to run the test. I don't have to run a bunch of `go build` commands before running tests.

How frequently do we actually want to test a specific, non-latest CLI version? Is it infrequent enough that we could support an optional `LINKERD_CLI` env to override it?

Reply:
Yea I similarly do not run the Go on non-latest CLI versions that often. Most of my interaction with using a non-latest CLI version is pulling with `docker-pull-cli-binaries` to install and test some specific issue, or pulling the binaries from a PR and checking the behavior.

I would assume passing in the binary to the integration tests is there only because the Go tests require it. If passing in the binary was removed from the Go tests, I don't really think we'd need to keep it around for the integration tests.

I wouldn't want this change to mean that every time tests are run it has to rebuild the binary to make sure it's up to date because that can introduce a lot of time, but continuing to use a binary for testing definitely keeps this RFC scoped to it's original purpose.

If there is a way to avoid that, I would be okay removing the CLI argument from the Go tests as well as the integration tests.

Reply:
How does that binary typically get built today? It used to be built in Docker, which seems overblown. Or, even if there is a good reason to build it in Docker, we can start using newer Docker features to support build caching... But if the test script had a `go build` (or even `docker build` with caching) in it, I'd expect that to be very fast when you're not touching CLI code.

Basically, what I'm getting at is.. if I'm working on a given Service and want to run an integration test against it, what does the dev/test loop look like? Ideally, we just run a script with the test target...

Reply:
The more I think about it, I think the current setup isn't so bad... but it would helpful to include some examples of the full test loop including whatever commands are typicaly needed to build the CLI to be able to run tests